scripts/dream.py

```diff
@@ -220,10 +222,70 @@ def main_loop(gen, opt, infile):
             prior_variations = opt.with_variations or []
             first_seed       = opt.seed
 
+            if opt.save_progress is not None or opt.show_progress is not None:
+                step_index = 1
+                step_count = 1
+
+            if opt.show_progress is not None:
+                if len(opt.show_progress) < 2:
+                    if len(opt.show_progress) == 0:
+                        opt.show_progress.extend([5, 2])
+                    elif len(opt.show_progress) == 1:
+                        opt.show_progress.append(2)
+                step_count = int(opt.show_progress[0])
+            
+            if opt.save_progress is not None:                                
+                if len(opt.save_progress) < 2:
+                    if len(opt.save_progress) == 0:
+                        opt.save_progress.extend([5, None])
+                    elif len(opt.save_progress) == 1:
+                        opt.save_progress.append(None)
+
+                step_count, progress_video_type = opt.save_progress
+            
+                if progress_video_type is not None:
+                    frames_for_video = []
+
+                if progress_video_type != 'vo':
+                    step_writer = PngWriter(os.path.join(current_outdir, 'intermediates'))
+                                
+            def image_progress(sample, step):
+                nonlocal step_index
+                nonlocal step_count
```

Comment:
nonlocals make for more complex code, generally good to be avoided. Options:
- you have access to opt so you could pull step_count of that
- you could use functool.partial(image_progress, a=a) and add another arg to image_process
whatever you think gets you simpler/shorter code with less potential for errors

Reply:
The nonlocals were from the initial PR by kevin that added the progress code. It is similar to how its done on the server.py model too. I'm not a fan of it either but I didn't change it for the time being.

Reply:
I'm with mh-dm on this one. `nonlocal` is ugly (almost as bad as GOTO 😂)

Unless I'm reading this wrong, this could even be a function outside of `main()` (still in 'dream.py'), with a (rather long) list of parameters.

Reply:
The issue here is that image_callback and step_callback are predefined callbacks that I cannot pass down any values too. And unfortunately the functionality of this and the image writer need it to use values from outside the scope of these functions. And I cannot initialize step_index inside the callback either because then it'll get initialized for iteration which of the steps.

With how dream.py is currently written, I don't see a way to bypass this. Probably why even lstein used nonlocals in image_writer.

Let me explore some options to see if there's a cleaner way for me to do this.